File: python/createMarket/create.market.test.py
#!/usr/bin/env python3
from dotenv import load_dotenv
import os
import base64
import base58
import requests
from solana.keypair import Keypair
from solana.transaction import VersionedTransaction

def main():
    # Load your PRIVATE_KEY from .env
    load_dotenv()
    private_key_b58 = os.getenv("PRIVATE_KEY")
    if not private_key_b58:
        raise EnvironmentError("Missing PRIVATE_KEY in environment")

    # Build your Solana wallet Keypair
    secret = base58.b58decode(private_key_b58)
    wallet = Keypair.from_secret_key(secret)
    print(f"wallet: {wallet.public_key}")

    # 1) Create market (prepare bundle of TXNs)
    create_url = "https://api.solanaportal.io/api/create/market"
    param = {
        "wallet_address": str(wallet.public_key),
        "mint": ""  # fill in your mint address
    }
    resp = requests.post(create_url, json=param)
    if resp.status_code != 200:
        print("Error creating market bundle:", resp.text)
        return

    data = resp.json()
    market_id = data.get("market_id")
    serialized_txns = data.get("serializedTxns", [])
    print(f"market id: {market_id}")

    # 2) Deserialize & sign each TXN, collect signatures and signed TXNs
    encoded_signed_txns = []
    signatures = []
    for tx_b64 in serialized_txns:
        txn_bytes = base64.b64decode(tx_b64)
        txn = VersionedTransaction.deserialize(txn_bytes)
        txn.sign([wallet])
        sig = base58.b58encode(txn.signatures[0]).decode()
        signatures.append(sig)
        signed_bytes = txn.serialize()
        encoded_signed_txns.append(base58.b58encode(signed_bytes).decode())

    # 3) Send the signed bundle via Jito
    jito_url = "https://tokyo.mainnet.block-engine.jito.wtf/api/v1/bundles"
    jito_payload = {
        "jsonrpc": "2.0",
        "id": 1,
        "method": "sendBundle",
        "params": [encoded_signed_txns],
    }
    jito_resp = requests.post(jito_url, json=jito_payload)
    if jito_resp.status_code == 200:
        for i, sig in enumerate(signatures):
            print(f"Transaction {i}: https://solscan.io/tx/{sig}")
    else:
        print("Bundle failed:", jito_resp.text)

if __name__ == "__main__":
    main()

File: typescript/createMarket/create.market.test.ts
import { Keypair, VersionedTransaction } from "@solana/web3.js";
import { bs58 } from "@project-serum/anchor/dist/cjs/utils/bytes";
import { configDotenv } from "dotenv";
configDotenv();
//send jito bundles transactions for token buy/sell

const pk = process.env.PRIVATE_KEY;
const test = async () => {
  try {
    const private_key = pk || "";
    const wallet = Keypair.fromSecretKey(bs58.decode(private_key));
    console.log("wallet:", wallet.publicKey.toBase58());

    const param = {
      wallet_address: wallet.publicKey.toBase58(),
      mint: "",
    };

    const url = "https://api.solanaportal.io/api/create/market";
    const response = await fetch(url, {
      method: "POST",
      headers: {
        "Content-Type": "application/json",
      },
      body: JSON.stringify(param),
    });

    // return;
    if (response.status === 200) {
      // successfully generated transaction
      const data = await response.json();
      const { market_id, serializedTxns } = data;
      console.log("market id:", market_id);
      const encodedSignedTxns = [];
      const signatures = [];
      for (let i = 0; i < serializedTxns.length; i++) {
        const txnBuffer = Buffer.from(serializedTxns[i], "base64");
        const txn = VersionedTransaction.deserialize(txnBuffer);
        txn.sign([wallet]);
        signatures.push(bs58.encode(txn.signatures[0]));
        const signedTxnBuffer = txn.serialize();
        encodedSignedTxns.push(bs58.encode(signedTxnBuffer));
      }

      try {
        const jitoResponse = await fetch(
          `https://tokyo.mainnet.block-engine.jito.wtf/api/v1/bundles`,
          {
            method: "POST",
            headers: {
              "Content-Type": "application/json",
            },
            body: JSON.stringify({
              jsonrpc: "2.0",
              id: 1,
              method: "sendBundle",
              params: [encodedSignedTxns],
            }),
          }
        );
        if (jitoResponse.status === 200) {
          for (let i = 0; i < signatures.length; i++) {
            console.log(
              `Transaction ${i}: https://solscan.io/tx/${signatures[i]}`
            );
          }
        } else {
          console.log(
            "bundle failed, please check the parameters",
            jitoResponse
          );
        }
      } catch (e: any) {
        console.error(e.message);
      }
    } else {
      console.log(response.statusText); // log error
    }
  } catch (e: any) {
    console.error(e.message);
  }
};

test();
